docs/modules/architecture/pages/compact-binary-specification.adoc

```diff
@@ -0,0 +1,719 @@
+|int16
+|SMALLINT
+|16-bit two's-complement signed integer
+|yes
+|int32
+|int
+|int32_t
+|int
+|int
+|number
+|int32
+|INTEGER
+|32-bit two's-complement signed integer
+|yes
+|int64
+|long
+|int64_t
+|long
+|int
+|Long
+|int64
+|BIGINT
+|64-bit two's-complement signed integer
+|yes
+|float32
+|float
+|float
+|float
+|float
+|number
+|float32
+|REAL
+|32-bit IEEE 754 floating-point number
+|yes
+|float64
+|double
+|double
+|double
+|float
+|number
+|float64
+|DOUBLE
+|64-bit IEEE 754 floating-point number
+|yes
+|string
+|String
+|optional<std::string>
+|string?
+|typing.Optional[str]
+|string \| null
+|*string
+|STRING
+|null or number of bytes in the string(iint32) + UTF-8 string https://tools.ietf.org/html/rfc3629
+|no
+|decimal
+|BigDecimal
+|optional<hazelcast:client:decimal>
+|HBigDecimal?
+|typing.Optional[decimal.Decimal]
+|BigDecimal \| null
+|*types.Decimal
+|DECIMAL
+|null or +
+Arbitrary precision and scale floating-point number: +
+represented as unscaledValue x 10 ^ -scale +
+unscaledValue: Array of int8 (byte array containing the two's-complement binary +
+representation in big-endian byte-order: the most significant byte is in the zeroth element.) +
+scale : single int32 for scale
+|no
+|time
+|LocalTime
+|optional<hazelcast:client:local_time
+|HLocalTime?
+|typing.Optional[datetime.time]
+|LocalTime \| null
+|*types.LocalTime
+|TIME
+|null or +
+HH-MI-SS-NN +
+int8: hour +
+int8: minute +
+int8: seconds +
+int32: nanoseconds +
+|no(since it is nullable)
+|date
+|LocalDate
+|optional<hazelcast:client:local_date>
+|HLocalDate?
+|typing.Optional[datetime.date]
+|LocalDate \| null
+|*types.LocalDate
+|DATE
+|null or +
+YYYY-MM-DD from -999999999-01-1 to 999999999-12-31 +
+int32: year +
+ int8: month +
+int8: dayOfMonth
+|no(since it is nullable)
+|timestamp
+|LocalDateTime
+|optional<hazelcast:client:local_date_time>
+|HLocalDateTime?
+|typing.Optional[datetime.datetime]
+|LocalDateTime \| null
+|*types.LocalDateTime
+|TIMESTAMP
+|null or +
+YYYY-MM-DD-HH-MI-SS-NN +
+int32: year +
+int8: month +
+int8: dayOfMonth +
+int8 : hour +
+int8: minute +
+int8: seconds +
+int32: nanoseconds +
+|no(since it is nullable)
+|timestampWithTimeZone
+|OffsetDateTime
+|optional<hazelcast:client:offset_date_time>
+|HOffsetDateTime?
+|typing.Optional[datetime.datetime]
+|OffsetDateTime \| null
+|*types.OffsetDateTime
+|TIMESTAMP W/ TZ
+|null or +
+YYYY-MM-DD-HH-MI-SS-MM Zone +
+int32: year +
+int8: month +
+int8:dayOfMonth +
+int8 : hour +
+int8: minute +
+int8: seconds +
+int32: nanoseconds +
+int32 : offsetSeconds. +
+offsetSeconds is range between +/-18:00:00 hour
+|no(since it is nullable)
+|compact
+|T
+|template<typename T> optional<T>
+|T?
+|typing.Optional[typing.Any]
+|T \| null
+|interface{}
+|OBJECT
+|A user defined compact
+|no
+|boolean[]
+|boolean[]
+|optional<std::vector<bool>>
+|bool[]?
+|typing.Optional[typing.List[bool]]
+|boolean[] \| null
+|[]bool
+|
+|Array of booleans
+|no
+|int8[]
+|byte[]
+|boost::optional<std::vector<int8_t>>
```

Comment:
sometimes we use `boost::optional` sometimes just the `optional`. I think we should stick to one for C++